File: java/org.financer.client/src/main/java/org/financer/client/connection/RequestConfig.java
package org.financer.client.connection;

import org.financer.shared.domain.model.api.DataTransferObject;
import org.financer.shared.path.Path;

import java.util.Collections;
import java.util.Map;

public class RequestConfig {

    private final HttpMethod httpMethod;
    private final String path;
    private final Map<String, String> queryParameters;
    private final DataTransferObject body;

    public RequestConfig(Path path) {
        this(path, Collections.emptyMap());
    }

    public RequestConfig(Path path, DataTransferObject body) {
        this(HttpMethod.fromString(path.getMethod()), path.getPath(), Collections.emptyMap(), body);
    }

    public RequestConfig(Path path, Map<String, String> parameters) {
        this(HttpMethod.fromString(path.getMethod()), path.getPath(), parameters, new DataTransferObject() {
            @Override
            public String toJson() {
                return "";
            }
        });
    }

    public RequestConfig(HttpMethod httpMethod, String path, Map<String, String> queryParameters, DataTransferObject body) {
        this.httpMethod = httpMethod;
        this.path = path;
        this.queryParameters = queryParameters;
        this.body = body;
    }

    public HttpMethod getHttpMethod() {
        return httpMethod;
    }

    public String getPath() {
        return path;
    }

    public Map<String, String> getQueryParameters() {
        return queryParameters;
    }

    public DataTransferObject getBody() {
        return body;
    }
}

File: java/org.financer.client/src/main/kotlin/org/financer/client/connection/RequestConfig.kt
package org.financer.client.connection

import org.financer.shared.domain.model.api.DataTransferObject
import org.financer.shared.path.Path

class RequestConfig(val httpMethod: HttpMethod, val path: String, val queryParameters: Map<String?, String?>, val body: DataTransferObject?) {

    constructor(path: Path, body: DataTransferObject?) : this(HttpMethod.Companion.fromString(path.method), path.path, emptyMap<String?, String>(), body) {}

    @JvmOverloads
    constructor(path: Path, parameters: Map<String?, String?> = emptyMap<String?, String>()) : this(HttpMethod.Companion.fromString(path.method), path.path, parameters, object : DataTransferObject {
        override fun toJson(): String {
            return ""
        }
    }) {
    }

}